internal/upload/pulp/pulp.go

```diff
@@ -178,3 +191,63 @@ func (cl *Client) TaskWaitingOrRunning(task string) bool {
 	}
 	return state == TASK_RUNNING || state == TASK_WAITING
 }
+
+// UploadAndDistributeCommit uploads a commit, creates a repository if
+// necessary, imports the commit to the repository, and distributes the
+// repository.
+func UploadAndDistributeCommit(archivePath, serverURL, repoName, basePath string, creds *Credentials) (string, error) {
+	client := NewClient(serverURL, creds)
+
+	// Check for the repository before uploading the commit:
+	// If the repository needs to be created but the basePath is empty, we
+	// should fail before uploading the commit.
+	log.Printf("checking if repository %q already exists", repoName)
+	repoHref, err := client.GetOSTreeRepositoryByName(repoName)
+	if err != nil {
+		return "", err
+	}
+
+	if repoHref == "" && basePath == "" {
+		return "", fmt.Errorf("repository %q does not exist and needs to be created, but no basepath for distribution was provided", repoName)
+	}
+
+	// Upload the file before creating the repository (if we need to create it)
+	// in case it fails. We don't want to have an empty repository if the
+	// commit upload fails.
+	log.Printf("uploading ostree commit to pulp server %s", serverURL)
+	fileHref, err := client.UploadFile(archivePath)
+	if err != nil {
+		return "", err
+	}
+
+	if repoHref == "" {
+		// repository does not exist: create it and distribute
+		log.Printf("repository not found - creating repository %q", repoName)
+		href, err := client.CreateOSTreeRepository(repoName, "")
+		if err != nil {
+			return "", err
+		}
+
+		repoHref = href
+		log.Printf("created repository %q (%s)", repoName, repoHref)
+		log.Printf("creating distribution at %q", basePath)
+		if _, err := client.DistributeOSTreeRepo(basePath, repoName, repoHref); err != nil {
+			return "", err
+		}
+	}
+
+	log.Printf("importing commit %q to repo %q", fileHref, repoHref)
+	if _, err := client.ImportCommit(fileHref, repoHref); err != nil {
+		return "", err
+	}
+
+	repoURL, err := url.JoinPath(serverURL, "pulp/content", basePath+"/")
+	if err != nil {
+		// this will happen if the serverURL is malformed, which would have
+		// already been caught by now
+		return "", err
+	}
+	log.Printf("repository url: %s", repoURL)
```

Comment:
I just realised that this is wrong.  The `basePath` can be empty or ignored if the repository already exists and has already been distributed.  We will have to find the distribution for the existing repository to return this.